dot/core/service.go

```diff
@@ -550,3 +554,61 @@ func (s *Service) GetMetadata(bhash *common.Hash) ([]byte, error) {
 	rt.SetContextStorage(ts)
 	return rt.Metadata()
 }
+
+// QueryStorage returns the key-value data by block based on `keys` params
+// on every block starting `from` until `to` block, if `to` is not nil
+func (s *Service) QueryStorage(from *common.Hash, to *common.Hash, keys ...string) (map[common.Hash]QueryKeyValueChanges, error) {
+	if from == nil {
+		return nil, errors.New("cannot query data without a starting block hash")
+	}
+
+	var err error
+	blocksToQuery := []common.Hash{*from}
+	if to != nil {
+		if blocksToQuery, err = s.blockState.SubChain(*from, *to); err != nil {
+			return nil, err
+		}
+	}
```

Comment:
I think if there is no `to` provided, you should set it to the `BestBlockHash`, the spec implies if no `to` is provided then you query every block from `to` until the head of the chain. can you double check what substrate returns?

Reply:
Ok! I will take a look at substrate node implementation

Reply:
Update to use the `BestBlockHash` if `to` is nil